oslo_vmware/image_transfer.py

```diff
@@ -346,3 +360,58 @@ def upload_image(context, timeout_secs, image_service, image_id, owner_id,
         updater.stop()
         read_handle.close()
     LOG.debug("Uploaded image: %s.", image_id)
+
+
+def image_pull_from_swift(swift_url, **kwargs):
+    (lease, lease_info) = \
+        rw_handles.VmdkWriteHandle._create_import_vapp_lease(
+            kwargs.get('session'), kwargs.get('resource_pool'),
+            kwargs.get('vm_import_spec'), kwargs.get('vm_folder'))
+
+    return image_pull_from_url(
+        kwargs.get('session'), lease, lease_info, swift_url.url,
+        size=kwargs.get('image_size'),
+        ssl_thumbprint=swift_url.ssl_thumbprint(),
+        headers={'X-Auth-Token': swift_url.auth_token})
+
+
+def image_pull_from_url(session, lease_ref, lease_info, url,
+                        size=None, ssl_thumbprint=None, headers=None):
+    """Converts a lease to pull from URLs and waits for the import.
+    Using HttpNfcLeasePullFromUrls_Task
+    """
+    if not lease_info.deviceUrl:
+        raise Exception("Invalid HttpNfc lease. "
+                        "No DeviceURLs found to import to.")
+
+    import_key = lease_info.deviceUrl[0].importKey
+
+    client_factory = session.vim.client.factory
+    file_spec = client_factory.create("ns0:HttpNfcLeaseSourceFile")
+    file_spec.targetDeviceId = import_key
+    file_spec.url = url
+    file_spec.create = True
+    file_spec.sslThumbprint = ssl_thumbprint
+    if size:
+        file_spec.size = size
+    if headers:
+        file_spec.httpHeaders = vim_util.dict_to_kv(client_factory,
+                                                    headers)
+    pull_task = session.invoke_api(session.vim,
+                                   "HttpNfcLeasePullFromUrls_Task",
+                                   lease_ref,
+                                   files=[file_spec])
+
+    LOG.debug("Started pulling the image from URL %(url)s",
+              {'url': url})
+
+    session.wait_for_task(pull_task)
+
+    LOG.debug("Completed pulling the image from URL %(url)s",
+              {'url': url})
+
+    session.invoke_api(session.vim,
```

Comment:
I think, the lease should be released in the same scope it gets acquired.
So, in `image_pull_from_swift`, also in a "finally"-block, ensuring it happens also when an exception occurs.

Reply:
Fixed.